src/compiler/mir/typetable.rs

```diff
@@ -0,0 +1,584 @@
+            Type::U32 => Self::U32,
+            Type::U64 => Self::U64,
+            Type::I8 => Self::I8,
+            Type::I16 => Self::I16,
+            Type::I32 => Self::I32,
+            Type::I64 => Self::I64,
+            Type::F64 => Self::F64,
+            Type::Bool => Self::Bool,
+            Type::StringLiteral => Self::StringLiteral,
+
+            _ => return None,
+        };
+        Some(base)
+    }
+}
+
+/// Defines a single type that exists within a Bramble program.
+#[derive(Debug, Clone)]
+pub enum MirTypeDef {
+    Base(MirBaseType),
+    Array { ty: TypeId, sz: usize },
+    RawPointer { mutable: PointerMut, target: TypeId },
+    Structure { path: Path, def: MirStructDef },
+}
+
+impl PartialEq for MirTypeDef {
+    /// Checks if two MirTypeDef values are logically the same type. Two [`MirTypeDef::Structure`] values
+    /// are equal if their [`Paths`](Path) are equal.
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Base(l0), Self::Base(r0)) => l0 == r0,
+            (Self::Array { ty: l_ty, sz: l_sz }, Self::Array { ty: r_ty, sz: r_sz }) => {
+                l_ty == r_ty && l_sz == r_sz
+            }
+            (
+                Self::RawPointer {
+                    mutable: l_mutable,
+                    target: l_target,
+                },
+                Self::RawPointer {
+                    mutable: r_mutable,
+                    target: r_target,
+                },
+            ) => l_mutable == r_mutable && l_target == r_target,
+            (Self::Structure { path: l_path, .. }, Self::Structure { path: r_path, .. }) => {
+                l_path == r_path
+            }
+            _ => false,
+        }
+    }
+}
+
+/// The Unique Identifier for a type within a Bramble program. Every type,
+/// including base types, is given a TypeId.  MIR uses the TypeId to annotate
+/// the types of MIR values and variables.
+#[derive(Debug, PartialEq, PartialOrd, Clone, Copy)]
+pub struct TypeId(u32);
+
+/// Provides the definition of a structure type. A structure may be referenced in another
+/// type before it is defined; in such a case, the structure is added to the type table
+/// as [`StructDef::Declared`].  When the structure is defined, the value will change to
+/// [`StructDef::Defined`], which will include the field definitions.
+#[derive(Debug, PartialEq, Clone)]
+pub enum MirStructDef {
+    /// A placeholder value that is used so that a TypeId can be given to a structure name
+    /// before the structure has been fully defined.
+    Declared,
+
+    /// The definition of a structure.
+    Defined(Vec<Field>),
+}
+
+impl MirStructDef {
+    /// If this structure has a matching field, then return the [`FieldId`].
+    /// Otherwise, return [`Option::None`].  This will also return [`Option::None`] if
+    /// this structure is in the [`MirStructDef::Declared`] state.
+    pub fn get_field_id(&self, name: StringId) -> Option<FieldId> {
```

Comment:
Should this return the `FieldId` and the field's type? I think when this is being used in MIR construction, we will want to fill in `FieldId` _and_ type of the field.

Reply:
Change name to `find_field`, because it is doing a search through the fields of the structure.